vizro-ai/src/vizro_ai/_vizro_ai.py

```diff
@@ -48,120 +38,41 @@ def __init__(self, model: Optional[Union[ChatOpenAI, str]] = None):
-        return self.components_instances[component_class]
-
-    def _run_plot_tasks(
-        self,
-        df: pd.DataFrame,
-        user_input: str,
-        max_debug_retry: int = 3,
-        explain: bool = False,
-        chart_name: Optional[str] = None,
-    ) -> PlotOutputs:
-        """Task execution."""
-        chart_type_pipeline = self.pipeline_manager.chart_type_pipeline
-        chart_type = chart_type_pipeline.run(initial_args={"chain_input": user_input, "df": df})
-
-        # TODO update to loop through charts for multiple charts creation
-        if chart_name is None:
-            chart_name = "custom_chart"
-        plot_pipeline = self.pipeline_manager.plot_pipeline
-        custom_chart_code = plot_pipeline.run(
-            initial_args={"chain_input": user_input, "df": df, "chart_type": chart_type, "chart_name": chart_name}
-        )
-
-        # TODO add debug in pipeline after getting _debug_helper logic in component
-        fix_func = self._lazy_get_component(GetDebugger).run
-        validated_code_dict = _debug_helper(
-            code_string=custom_chart_code, max_debug_retry=max_debug_retry, fix_chain=fix_func, df=df
-        )
-
-        pass_validation = validated_code_dict.get("debug_status")
-        code_string = validated_code_dict.get("code_string")
-
-        if not pass_validation:
-            raise DebugFailure(
-                "Chart creation failed. Retry debugging has reached maximum limit. Try to rephrase the prompt, "
-                "or try to select a different model. Fallout response is provided: \n\n" + code_string
-            )
-
-        fig_object = _exec_code_and_retrieve_fig(code=code_string, local_args={"df": df}, is_notebook_env=_is_jupyter())
-        if explain:
-            business_insights, code_explanation = self._lazy_get_component(GetCodeExplanation).run(
-                chain_input=user_input, code_snippet=code_string
-            )
-
-            return PlotOutputs(
-                code=code_string,
-                figure=fig_object,
-                business_insights=business_insights,
-                code_explanation=code_explanation,
-            )
-
-        return PlotOutputs(code=code_string, figure=fig_object)
-
-    def _get_chart_code(self, df: pd.DataFrame, user_input: str) -> str:
-        """Get Chart code of vizro via english descriptions, English to chart translation.
-
-        Can be used in integration with other application if only code snippet return is required.
-
-        Args:
-            df: The dataframe to be analyzed
-            user_input: User questions or descriptions of the desired visual
-
-        """
-        # TODO retained for some chat application integration, need deprecation handling
-        return self._run_plot_tasks(df, user_input, explain=False).code
 
     def plot(
         self,
         df: pd.DataFrame,
         user_input: str,
-        explain: bool = False,
         max_debug_retry: int = 3,
```

Comment:
Did you try estimate the optimal retry number? I understand retry is helpful as it can feedback the error message to llm. But wonder whether retry 3 times is more effective than 2 times? Do we observe llm converge to correct answers with more retries, or llm is directed to a direction of off-track?

Unless 3 is obviously better, I'd prefer maybe even retry once. Because it 3 means triple the latency, which is long